openeo/extra/job_management.py

```diff
@@ -0,0 +1,290 @@
+
+            time.sleep(self.poll_sleep)
+
+    def on_job_done(self, job: BatchJob, row):
+        """
+        Handles jobs that have finished. Can be overridden to provide custom behaviour.
+
+        Default implementation downloads the results into a folder containing the title.
+
+        :param job: The job that has finished.
+        :param row: DataFrame row containing the job's metadata.
+        """
+        # TODO: param `row` is never accessed in this method. Remove it? Is this intended for future use?
+
+        job_metadata = job.describe_job()
+        job.get_results().download_files(target=job_metadata["title"])
+        with open(Path(job_metadata["title"]) / f"job_{job.job_id}.json", "w") as f:
+            json.dump(job_metadata, f, ensure_ascii=False)
+
+    def on_job_error(self, job: BatchJob, row):
+        """
+        Handles jobs that stopped with errors.  Can be overridden to provide custom behaviour.
+
+        Default implementation writes the error logs to a JSON file.
+
+        :param job: The job that has finished.
+        :param row: DataFrame row containing the job's metadata.
+        """
+        # TODO: param `row` is never accessed in this method. Remove it? Is this intended for future use?
+
+        logs = job.logs()
+        error_logs = [l for l in logs if l.level.lower() == "error"]
+        job_metadata = job.describe_job()
+
+        title = job_metadata["title"]
+        if len(error_logs) > 0:
+            f"job_{title}_errors.json".write_text(json.dumps(error_logs, indent=2))
+
+    def _update_statuses(self, df: pd.DataFrame):
+        """Update status (and stats) of running jobs (in place)."""
+        active = df.loc[
+            (df.status == "created")
+            | (df.status == "queued")
+            | (df.status == "running")
+        ]
+        for i in active.index:
+            job_id = df.loc[i, "id"]
+            backend_name = df.loc[i, "backend_name"]
+
+            try:
+                con = self.backends[backend_name].get_connection()
+                the_job = con.job(job_id)
+                job_metadata = the_job.describe_job()
+                _log.info(
+                    f"Status of job {job_id!r} (on backend {backend_name}) is {job_metadata['status']!r}"
+                )
+                if (
+                    df.loc[i, "status"] == "running"
+                    and job_metadata["status"] == "finished"
+                ):
+                    self.on_job_done(the_job, df.loc[i])
+                if df.loc[i, "status"] != "error" and job_metadata["status"] == "error":
+                    self.on_job_error(the_job, df.loc[i])
+
+                df.loc[i, "status"] = job_metadata["status"]
+                for key in job_metadata.get("usage", {}).keys():
+                    df.loc[i, key] = _format_usage_stat(job_metadata, key)
+
+            except OpenEoApiError as e:
+                print(f"error for {backend_name}")
```

Comment:
This error printing should also contain `i`, or `job_id` I think to be more useful

Reply:
Added the job_id to the f-string.